Fails to build for target `wasm32-unknown-unknown`

Hey there, I'd love to use this library for in-browser crunch decompression, but unfortunately it doesn't compile for `wasm32-unknown-unknown`:

```
[I] ~/d/crnlib [master] ➜ cargo build --target wasm32-unknown-unknown
   Compiling proc-macro2 v1.0.93
   Compiling unicode-ident v1.0.14
   Compiling serde v1.0.217
   Compiling anyhow v1.0.95
   Compiling either v1.13.0
   Compiling radium v0.3.0
   Compiling bitvec v0.17.4
   Compiling quote v1.0.38
   Compiling syn v2.0.96
   Compiling serde_derive v1.0.217
   Compiling serde_repr v0.1.19
   Compiling bincode v1.3.3
   Compiling crnlib v0.1.1 (/Users/enoch/dev/crnlib)
error[E0277]: the trait bound `u64: BitStore` is not satisfied
   --> src/codec.rs:18:43
    |
18  |       self.buffer[self.index..].load_be::<u64>() << (self.index + n - self.buffer.len())
    |                                 -------   ^^^ the trait `BitStore` is not implemented for `u64`
    |                                 |
    |                                 required by a bound introduced by this call
    |
    = help: the following other types implement trait `BitStore`:
              u16
              u32
              u8
              usize
note: required by a bound in `load_be`
   --> /Users/enoch/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bitvec-0.17.4/src/fields.rs:173:11
    |
172 |     fn load_be<U>(&self) -> U
    |        ------- required by a bound in this associated function
173 |     where U: BitStore;
    |              ^^^^^^^^ required by this bound in `BitField::load_be`

error[E0277]: the trait bound `u64: BitStore` is not satisfied
   --> src/codec.rs:20:45
    |
20  |       self.buffer[self.index..self.index+n].load_be()
    |                                             ^^^^^^^ the trait `BitStore` is not implemented for `u64`
    |
    = help: the following other types implement trait `BitStore`:
              u16
              u32
              u8
              usize
note: required by a bound in `load_be`
   --> /Users/enoch/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bitvec-0.17.4/src/fields.rs:173:11
    |
172 |     fn load_be<U>(&self) -> U
    |        ------- required by a bound in this associated function
173 |     where U: BitStore;
    |              ^^^^^^^^ required by this bound in `BitField::load_be`

error[E0277]: the trait bound `u64: BitStore` is not satisfied
   --> src/codec.rs:29:56
    |
29  |     let result = self.buffer[self.index..self.index+n].load_be();
    |                                                        ^^^^^^^ the trait `BitStore` is not implemented for `u64`
    |
    = help: the following other types implement trait `BitStore`:
              u16
              u32
              u8
              usize
note: required by a bound in `load_be`
   --> /Users/enoch/.cargo/registry/src/index.crates.io-6f17d22bba15001f/bitvec-0.17.4/src/fields.rs:173:11
    |
172 |     fn load_be<U>(&self) -> U
    |        ------- required by a bound in this associated function
173 |     where U: BitStore;
    |              ^^^^^^^^ required by this bound in `BitField::load_be`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `crnlib` (lib) due to 3 previous errors
```

I'd be happy to work on these if you're interested in mainlining this kind of feature.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "crnlib"
-version = "0.1.1"
+version = "0.1.2"
 authors = ["Clouds Flowing <clouds.flowing@gmail.com>"]
 edition = "2018"
 repository = "https://github.com/clouds56/crnlib/"

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -18,4 +18,5 @@ bitvec = "0.17"
 anyhow = "1.0"
 
 [dev-dependencies]
-image = ">= 0.23"
+# see https://github.com/image-rs/image/issues/2404
+image = "= 0.23"

--- a/src/codec.rs
+++ b/src/codec.rs
@@ -1,5 +1,5 @@
 use std::collections::BTreeMap;
-use bitvec::{slice::BitSlice, order::Msb0, fields::BitField};
+use bitvec::{slice::BitSlice, store::BitStore, order::Msb0, fields::BitField};
 use anyhow::*;
 
 pub struct Codec<'a> {
@@ -11,21 +11,21 @@ impl Codec<'_> {
   pub fn new<'a>(input: &'a [u8]) -> Codec<'a> {
     Codec { buffer: BitSlice::from_slice(input), index: 0 }
   }
-  pub fn look_bits(&self, n: usize) -> u64 {
-    assert!(n <= 64);
-    if n == 0 { return 0 }
+  pub fn look_bits<T: BitStore>(&self, n: usize) -> T {
+    assert!(n <= T::BITS as usize);
+    if n == 0 { return T::FALSE }
     if self.index + n > self.buffer.len() {
-      self.buffer[self.index..].load_be::<u64>() << (self.index + n - self.buffer.len())
+      self.buffer[self.index..].load_be::<T>() << (self.index + n - self.buffer.len()) as u8
     } else {
       self.buffer[self.index..self.index+n].load_be()
     }
   }
-  pub fn read_bits(&mut self, n: usize) -> Result<u64, Error> {
-    assert!(n <= 64);
+  pub fn read_bits<T: BitStore>(&mut self, n: usize) -> Result<T, Error> {
+    assert!(n <= T::BITS as usize);
     if self.index + n > self.buffer.len() {
       bail!("read out of index {} < {}", self.index+n, self.buffer.len());
     }
-    if n == 0 { return Ok(0) }
+    if n == 0 { return Ok(T::FALSE) }
     let result = self.buffer[self.index..self.index+n].load_be();
     self.index += n;
     Ok(result)
@@ -44,17 +44,17 @@ impl Codec<'_> {
   }
 
   pub fn get_huffman(&mut self) -> Result<Huffman<u32>, Error> {
-    let symbol_count = self.read_bits(Huffman::<()>::MAX_SYMBOL_COUNT_BIT)? as u32;
+    let symbol_count = self.read_bits::<u32>(Huffman::<u32>::MAX_SYMBOL_COUNT_BIT)?;
     // println!("construct huffman tree with {} symbols", symbol_count);
     if symbol_count == 0 {
       return Huffman::new(BTreeMap::new())
     }
     let mut tmp_symbol_depth = BTreeMap::new();
-    let tmp_symbol_count = self.read_bits(5)? as usize;
+    let tmp_symbol_count = self.read_bits::<usize>(5)?;
     ensure!(tmp_symbol_count <= Key::SHUFFLE.len(),
       "tmp_symbol_count {} > {}", tmp_symbol_count, Key::SHUFFLE.len());
     for i in 0..tmp_symbol_count {
-      let value = self.read_bits(3)? as usize;
+      let value = self.read_bits::<usize>(3)?;
       if value != 0 {
         tmp_symbol_depth.insert(Key::SHUFFLE[i], value);
       }
@@ -66,11 +66,11 @@ impl Codec<'_> {
     let mut last = None;
     while i < symbol_count {
       let (len, d) = match key.next(self).context("get key content")? {
-        Depth(d) => (1, d),
-        ShortZero => (self.read_bits(3)? + 3, 0),
-        LongZero => (self.read_bits(7)? + 11, 0),
-        ShortRepeat => (self.read_bits(2)? + 3, last.ok_or_else(|| anyhow!("short repeat no last"))?),
-        LongRepeat => (self.read_bits(6)? + 7, last.ok_or_else(|| anyhow!("long repeat no last"))?),
+        Depth(d) => (1u32, d),
+        ShortZero => (self.read_bits::<u32>(3)? + 3, 0),
+        LongZero => (self.read_bits::<u32>(7)? + 11, 0),
+        ShortRepeat => (self.read_bits::<u32>(2)? + 3, last.ok_or_else(|| anyhow!("short repeat no last"))?),
+        LongRepeat => (self.read_bits::<u32>(6)? + 7, last.ok_or_else(|| anyhow!("long repeat no last"))?),
       };
       last = Some(d);
       for j in 0..len as u32 {
@@ -89,11 +89,11 @@ impl Codec<'_> {
 fn test_read_bits() {
   let input = [0b1100_1010u8, 0b0110_1101, 0b1101_1001];
   let mut codec = Codec::new(&input);
-  assert_eq!(codec.read_bits(3).unwrap(), 0b110);
+  assert_eq!(codec.read_bits::<u8>(3).unwrap(), 0b110);
   assert_eq!(codec.index, 3);
-  assert_eq!(codec.read_bits(17).unwrap(), 0b1010_0110_1101_1101);
+  assert_eq!(codec.read_bits::<u32>(17).unwrap(), 0b1010_0110_1101_1101);
   assert_eq!(codec.index, 20);
-  assert_eq!(codec.read_bits(0).unwrap(), 0);
+  assert_eq!(codec.read_bits::<u8>(0).unwrap(), 0);
   assert_eq!(codec.index, 20);
 
   assert_eq!(Huffman::<()>::MAX_SYMBOL_COUNT, 1 << (Huffman::<()>::MAX_SYMBOL_COUNT_BIT - 1));
@@ -164,12 +164,14 @@ impl<T: Ord+Copy> Huffman<T> {
       symbols, symbol_rev,
     })
   }
+}
 
+impl<T: Ord+Copy> Huffman<T> {
   pub fn next(&self, codec: &mut Codec<'_>) -> Result<T, Error> {
     ensure!(codec.current() < codec.len(), "stream end {} >= {}", codec.current(), codec.len());
-    let k = codec.look_bits(self.max_depth) as u32;
+    let k = codec.look_bits::<u32>(self.max_depth);
     for i in 1..=self.max_depth {
-      let t = k >> (self.max_depth - i);
+      let t = k >> (self.max_depth - i) as u8;
       if let Some(sym) = self.symbol_rev.get(&(i, t)) {
         codec.index += i;
         return Ok(*sym)

